make JALR more readable (#880)
riscv/riscv-cheri: src/cheri/insns/jalr_32bit.adoc

#Begin changed since last ARC review#

Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions and/or encoding formats to give a large degree of extensibility.

>  Unless an extension and/or encoding format, such as <<section_rvy_sentry_insn_ext>>,  is implemented which explicitly reference the hooks, then take no action for any of them.

Reviewer: ```suggestion
NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions to give a large degree of extensibility.
 Unless an extension, such as <<section_rvy_sentry_insn_ext>>, is implemented that explicitly reference the hooks, then take no action for any of them.
```

I thought the conclusion was to avoid mentioning encodings here, since we are only discussing semantics here. @nwf does the suggested wording sound good to you?

Author: it's good for me

Reviewer 2: A challenge to that is that the default encoding chapters also include semantics, so it's hard to talk about the one without the other.

+
Operationally, copy `{cs1}` to the target <<pcc>> and then...
+
. Compute the target <<pcc>> address in two steps, each using the semantics of the <<SCADDR>> instruction:
.. Increment the address of the target <<pcc>> by the sign-extended 12-bit `offset`, and
.. set bit zero of the target <<pcc>> address to zero.

> . HOOK 1: Optionally clear the {ctag} of the target <<pcc>>.

Reviewer: ```suggestion
. HOOK 1: Optionally clear the {ctag} of the target <<pcc>> depending on the <<sec_cap_type>> value and the numerical values of `{cd}` and `{cs1}`
```

> . HOOK 1: Optionally clear the {ctag} of the target <<pcc>>.

Reviewer: Maybe something like that

. HOOK 2: Optionally unseal the target <<pcc>> if sealed.
. If the target <<pcc>> is (still) sealed, clear its {ctag}.
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:

> .. add the width of this instruction to the current <<pcc>>, and

Reviewer 2: While here, this should be amended to say "using the semantics of the `<<SCADDR>>` instruction" since the addition might, in principle, take us out of representable bounds and we need to make sure that we don't construct an incorrectly reinterpreted capability.

> .. HOOK 3: Optionally seal the return capability.

Author: how would you specify it?

#Begin changed since last ARC review#

Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+

> NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions and/or encoding formats to give a large degree of extensibility.

Reviewer: ```suggestion
NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions to give a large degree of extensibility.
```

 The simpler operation for <<rv32y_cap_description>>/<<rv64y_cap_description>> only uses hook 2.
 Examples of how CHERIoT encoding formats uses all of the hooks is included below the description.
 See the full details of the CHERIoT encoding formats for the full specifications, for example: <<rv32y_cheriot_encoding1_description>>.
+
Operationally, copy `{cs1}` to the target <<pcc>> and then...
+
. Compute the target <<pcc>> address in two steps, each using the semantics of the <<SCADDR>> instruction:
.. Increment the address of the target <<pcc>> by the sign-extended 12-bit `offset`, and
.. set bit zero of the target <<pcc>> address to zero.
. HOOK 1: Clear the {ctag} of the target <<pcc>> if the triple of its <<sec_cap_type>> value, `{cd}`, and `{cs1}` is not acceptable, as defined by the capability encoding.
. If the target <<pcc>> is a <<sentry_cap>>, then unseal it.
. If the target <<pcc>> is (still) sealed, clear its {ctag}.
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>>, and

> .. HOOK 2: if the capability encoding defines <<sentry_cap>> <<sec_cap_type>> values for return pointers, then seal the return capability as specified.

Reviewer 2: ```suggestion
.. HOOK 2: if the capability encoding or an extension defines <<sentry_cap>> <<sec_cap_type>> values for return pointers, then seal the return capability as specified.
```

Reviewer: I wonder if this should be extension again? Then it could be dependent on Zys/Zycheriot/Zycfi?

+
. Compute the target <<pcc>> address in two steps, each using the semantics of the <<SCADDR>> instruction:
.. Increment the address of the target <<pcc>> by the sign-extended 12-bit `offset`, and
.. set bit zero of the target <<pcc>> address to zero.
. HOOK 1: Clear the {ctag} of the target <<pcc>> if the triple of its <<sec_cap_type>> value, `{cd}`, and `{cs1}` is not acceptable, as defined by the capability encoding.
. If the target <<pcc>> is a <<sentry_cap>>, then unseal it.
. If the target <<pcc>> is (still) sealed, clear its {ctag}.
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>>, and
.. HOOK 2: if the capability encoding defines <<sentry_cap>> <<sec_cap_type>> values for return pointers, then seal the return capability as specified.
. Jump to the target <<pcc>>.
.. NOTE: It is not possible for the target <<pcc>> to be sealed by the time it is jumped to.
. HOOK 3: Apply any capability encoding specified architectural semantics for the type of `{cs1}`.
+
The above procedure for <<JALR_CHERI>> defines three hooks for use by the capability encoding:
+

> 1. Conditional acceptability of the triple of the two _register selectors_ (`{cd}`, `{cs1}`) used in an instance of the instruction together with the <<sec_cap_type>> value of the capability in `{cs1}`.

Author: maybe - but that was a lot of the problem with the text before - there was so much extra information that it was hard to see the actual procedure to follow
I think I've done what I can with this one - I find it clearer than before.
